hydra-plutus/src/Hydra/Contract/Head.hs

```diff
@@ -248,46 +257,41 @@ checkClose ::
   ScriptContext ->
   [Party] ->
   BuiltinByteString ->
-  SnapshotNumber ->
-  BuiltinByteString ->
   [Signature] ->
   ContestationPeriod ->
   CurrencySymbol ->
   Bool
-checkClose ctx parties initialUtxoHash snapshotNumber closedUtxoHash sig cperiod headPolicyId =
+checkClose ctx parties initialUtxoHash sig cperiod headPolicyId =
   hasBoundedValidity
+    && checkDeadline
     && checkSnapshot
     && mustBeSignedByParticipant ctx headPolicyId
     && hasST headPolicyId outValue
  where
-  hasBoundedValidity = traceIfFalse "hasBoundedValidity check failed" $ tMax - tMin <= cp
+  hasBoundedValidity =
+    traceIfFalse "hasBoundedValidity check failed" $
+      tMax - tMin <= cp
 
   outValue =
     maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
 
+  (closedSnapshotNumber, closedUtxoHash, closedContestationDeadline) =
+    -- XXX: fromBuiltinData is super big (and also expensive?)
+    case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
+      Just Closed{snapshotNumber, utxoHash, contestationDeadline} -> (snapshotNumber, utxoHash, contestationDeadline)
+      _ -> traceError "wrong state in output datum"
+
   checkSnapshot
-    | snapshotNumber == 0 =
-      let expectedOutputDatum =
-            Closed
-              { parties
-              , snapshotNumber = 0
-              , utxoHash = initialUtxoHash
-              , contestationDeadline = makeContestationDeadline cperiod ctx
-              , headId = headPolicyId
-              }
-       in checkHeadOutputDatum ctx expectedOutputDatum
```

Comment:
This was also checking that the `headId` and `parties` don't change.

Should: do the same as `mustNotChangeParameters` below.

(I added `mustNotChangeParameters` to see whether we would run into bigger scripts / costs if we completely rewrite the checks that was there before.)

Reply:
headId is checked line 269:

```
    && hasST headPolicyId outValue
```

Reply:
Added two new mutations to mess around with the headId and the parties to exhibit the issue and fixed. See last commit.